packages/vite-extensions/vite-preserve-loaders-imports.ts

```
@@ -0,0 +1,58 @@
+import type { Plugin } from 'vite';
+
+export interface PreserveLoadersRule {
+	regex: RegExp;
+	transform: (specifier: string) => string;
+}
+
+export function vitePreserveLoadersImports(
+	rules: PreserveLoadersRule[]
+): Plugin {
+	let command: 'build' | 'serve';
+
+	return {
+		/**
+		 * Vite can't extract static asset in the library mode:
+		 * https://github.com/vitejs/vite/issues/3295
+		 *
+		 * This workaround replaces the actual php.js modules paths used
+		 * in the dev mode with their filenames. Then, the filenames are marked
+		 * as external further down in this config. As a result, the final
+		 * bundle contains literal `import('php.js')` and
+		 * `import('php.wasm')` statements which allows the consumers to use
+		 * their own loaders.
+		 *
+		 * This keeps the dev mode working AND avoids inlining 5mb of
+		 * wasm via base64 in the final bundle.
+		 */
+		name: 'vite-preserve-loaders-imports',
+
+		configResolved(config) {
+			command = config.command;
+		},
+
+		resolveDynamicImport(specifier) {
+			if (command !== 'build' || typeof specifier !== 'string') return;
+
+			for (const rule of rules) {
+				if (rule.regex.test(specifier)) {
```

What would it take to introduce a poka-yoke and throw an error when a rule matches zero specifiers?

---

We could add a  `const matchedRules = new Set<PreserveLoadersRule>();` and add a matching rule :

```javascript
if (rule.regex.test(specifier)) {
	matchedRules.add(rule);
...
```

And during `buildEnd` we could throw an error if a rule has not been specified : 

```
buildEnd() {
	if (command !== 'build') return;

	const unusedRules = rules.filter(rule => !matchedRules.has(rule));

	if (unusedRules.length > 0) {
		const details = unusedRules
			.map((rule) => `- ${rule.regex}`)
			.join('\n');

		this.error(
			`vite-preserve-loaders-imports: The following rules did not match any dynamic imports:\n${details}\n\n` +
			`This is likely a misconfiguration or a stale regex.`
		);
	}
},
```

But I think if some regex were not correct Vite would crash anyway.

---

It probably would! Still, having an explicit line of defense is always reassuring – we can stay resilient even if something implicit changes and we never realize. Let's add that buildEnd() logic.

---

done! Thank you Adam.